File: benchmarks/livermore/k19_linear_recurrence.c
/* K19 — General Linear Recurrence Equations (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION B5(1001), SA(1001), SB(1001), N=101
   Forward: DO K=1,N: B5(K+KB5I) = SA(K) + STB5*SB(K); STB5 = B5(K+KB5I) - STB5
   Backward: DO I=1,N: K=N-I+1; same recurrence */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     101
#define NREPS 10000

int main(void) {
    double b5[1002], sa[1002], sb[1002];

    signel(sa+1, 101);
    signel(sb+1, 101);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    long kb5i;
    for (long rep = 0; rep < NREPS; rep++) {
        double stb5 = 0.5;
        kb5i = 0;
        /* Forward sweep */
        for (long k = 1; k <= N; k++) {
            b5[k+kb5i] = sa[k] + stb5 * sb[k];
            stb5 = b5[k+kb5i] - stb5;
        }
        /* Backward sweep */
        for (long i = 1; i <= N; i++) {
            long k = N - i + 1;
            b5[k+kb5i] = sa[k] + stb5 * sb[k];
            stb5 = b5[k+kb5i] - stb5;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("b5[51] = %f\n", b5[51]);
    return 0;
}

File: benchmarks/livermore/k19_linear_recurrence.f
C     Kernel 19 -- General Linear Recurrence Equations
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K19
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION B5(1001), SA(1001), SB(1001)
C
C     Initialize SA using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      SCALED = 0.1D0
      DO 10 K = 1, 101
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         SA(K) = (BUZZ - FIZZ) * SCALED
   10 CONTINUE
C
C     Initialize SB using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 15 K = 1, 101
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         SB(K) = (BUZZ - FIZZ) * SCALED
   15 CONTINUE
C
C     Zero out B5
C
      DO 20 K = 1, 1001
         B5(K) = 0.0D0
   20 CONTINUE
C
      N = 101
C
C     Kernel 19: General Linear Recurrence Equations
C     Forward and backward sweeps with recurrence
C
      DO 500 IREP = 1, 10000
         STB5 = 0.5D0
         KB5I = 0
C
C        Forward sweep
         DO 200 K = 1, N
            B5(K+KB5I) = SA(K) + STB5*SB(K)
            STB5 = B5(K+KB5I) - STB5
  200    CONTINUE
C
C        Backward sweep
         DO 300 I = 1, N
            K = N - I + 1
            B5(K+KB5I) = SA(K) + STB5*SB(K)
            STB5 = B5(K+KB5I) - STB5
  300    CONTINUE
  500 CONTINUE
C
      WRITE(*,900) B5(51)
  900 FORMAT('K19 linear recurrence: b5(51) = ', E23.15E3)
      END
